File: ios/RunWrap/CollectionEngine.swift
import Foundation

/// 도감에 수집되는 새 종 — 사이클 목표 수준이 종을 정한다 (기획서 §5).
///
/// 매핑은 초안이고 **"서브3 = 백조"만 확정 축**이다. 나머지 경계는 에셋 수급과 함께
/// 확정되므로(§12), 경계값을 `CollectionEngine.species(for:)` 한 곳에만 두고
/// 화면·저장 어디에서도 다시 판정하지 않는다.
enum BirdSpecies: String, Codable, CaseIterable {
    case sparrow    // 참새
    case swallow    // 제비
    case falcon     // 매
    case goose      // 기러기
    case crane      // 두루미
    case swan       // 백조

    var label: String {
        switch self {
        case .sparrow: "참새"
        case .swallow: "제비"
        case .falcon: "매"
        case .goose: "기러기"
        case .crane: "두루미"
        case .swan: "백조"
        }
    }

    /// 도감에서 이 종이 어떤 목표의 증표인지 한 줄로 설명한다 (미수집 칸의 힌트로도 쓴다)
    var goalHint: String {
        switch self {
        case .sparrow: "목표 없이 완주 습관"
        case .swallow: "5K · 10K"
        case .falcon: "하프"
        case .goose: "풀코스 완주"
        case .crane: "풀코스 sub-4"
        case .swan: "풀코스 서브3"
        }
    }
}

/// 도감에 수록된 새 한 마리 — 성조 이미지 + 당시 목표 + 수집일 (기획서 §5)
struct CollectedBird: Codable, Equatable, Identifiable {
    let id: UUID
    let species: BirdSpecies
    /// 수집 시점의 목표를 사람이 읽는 문장으로 굳혀 둔다.
    /// 목표 설정은 나중에 바뀌므로 rawValue가 아니라 **그때의 표기**를 남긴다 —
    /// 도감은 이력이라 과거 항목이 현재 설정을 따라 바뀌면 안 된다.
    let goalLabel: String
    let collectedAt: Date
    /// 이 사이클이 걸린 일수 — 도감에서 "27일 만에" 같은 이력 표기에 쓴다
    let cycleDays: Int

    init(id: UUID = UUID(), species: BirdSpecies, goalLabel: String,
         collectedAt: Date, cycleDays: Int) {
        self.id = id
        self.species = species
        self.goalLabel = goalLabel
        self.collectedAt = collectedAt
        self.cycleDays = cycleDays
    }
}

/// 성조 도달 → 수집 → 새 사이클 시작을 계산하는 순수 로직 (기획서 §5).
///
/// 엔진 계층 규칙 그대로 Foundation만 import하고 `now`를 주입받는다.
/// 저장·화면 전환은 `CollectionStore`와 화면이 맡고, 여기서는 **무엇을 수집하고
/// 다음 사이클을 어떤 값으로 시작할지**만 정한다.
enum CollectionEngine {
    /// 서브3 기준 — 풀코스 3시간. "서브3 = 백조"가 §5의 확정 축이다
    private static let sub3Seconds = 3 * 3_600
    /// sub-4 기준 — 풀코스 4시간. 두루미와 기러기의 경계
    private static let sub4Seconds = 4 * 3_600

    /// 사이클 목표 → 새 종 (기획서 §5 매핑표).
    ///
    /// - Parameters:
    ///   - distance: 그 사이클의 목표 종목. nil이면 목표 없음(완주 습관) → 참새
    ///   - goalSeconds: 목표 기록(초). 0 이하는 미입력으로 본다 — 풀코스에서만 종을 가른다
    static func species(for distance: RaceDistance?, goalSeconds: Int) -> BirdSpecies {
        guard let distance else { return .sparrow }
        switch distance {
        case .fiveK, .tenK: return .swallow
        case .half: return .falcon
        case .full:
            // 목표 기록 미입력이면 "완주"로 본다 — 기러기.
            // 기록을 적어 낸 경우에만 sub-4 / 서브3로 승격한다
            guard goalSeconds > 0 else { return .goose }
            if goalSeconds < sub3Seconds { return .swan }
            if goalSeconds < sub4Seconds { return .crane }
            return .goose
        }
    }

    /// 목표를 도감에 남길 문장으로 굳힌다 — 종목 + (있으면) 목표 기록.
    /// 예: "풀코스 3:30:00" · "하프" · "목표 없이 완주 습관"
    static func goalLabel(for distance: RaceDistance?, goalSeconds: Int) -> String {
        guard let distance else { return BirdSpecies.sparrow.goalHint }
        guard goalSeconds > 0 else { return distance.label }
        return "\(distance.label) \(Format.duration(Double(goalSeconds)))"
    }

    /// 성조에 도달했는지 — 수집 세러모니를 띄울 조건.
    ///
    /// 미노출 가드와 같은 결의 판정이다: 단계가 실제로 `flying`일 때만 true다.
    static func hasReachedAdult(stage: GrowthStage) -> Bool {
        stage.next == nil
    }

    /// 수집할 새를 만든다. 사이클 시작 시각과 지금으로 소요 일수를 함께 굳힌다.
    static func collect(distance: RaceDistance?, goalSeconds: Int,
                        cycleStartedAt: Date, now: Date,
                        id: UUID = UUID()) -> CollectedBird {
        var calendar = Calendar(identifier: .iso8601)
        calendar.timeZone = .current
        let days = calendar.dateComponents([.day],
                                            from: calendar.startOfDay(for: cycleStartedAt),
                                            to: calendar.startOfDay(for: now)).day ?? 0
        return CollectedBird(id: id,
                             species: species(for: distance, goalSeconds: goalSeconds),
                             goalLabel: goalLabel(for: distance, goalSeconds: goalSeconds),
                             collectedAt: now,
                             cycleDays: max(0, days))
    }

    /// 다음 목표 추천 — 직전 목표를 한 칸 올린다 (기획서 §5 "직전 목표·최근 기록 기반").
    ///
    /// 풀코스에 이미 도달한 뒤에는 종목을 더 올릴 수 없으므로 **기록 단축**으로 방향을 튼다.
    /// 추천일 뿐이라 화면에서 직접 설정으로 바꿀 수 있다.
    ///
    /// - Returns: 추천 종목과 추천 목표 기록(초, 0이면 기록 목표 없이 완주). 더 올릴 곳이
    ///   없으면 nil — 그 경우 화면은 추천 대신 직접 설정만 안내한다
    static func recommendedGoal(after distance: RaceDistance?,
                                 goalSeconds: Int) -> (distance: RaceDistance, seconds: Int)? {
        switch distance {
        case .none: return (.fiveK, 0)
        case .fiveK: return (.tenK, 0)
        case .tenK: return (.half, 0)
        case .half: return (.full, 0)
        case .full:
            // 종목은 끝 — 기록을 당긴다. 미입력이면 우선 sub-4를 제안하고,
            // 이미 목표가 있으면 30분씩 당기되 서브3 아래로는 내리지 않는다.
            guard goalSeconds > 0 else { return (.full, sub4Seconds) }
            let tightened = goalSeconds - 30 * 60
            guard tightened >= sub3Seconds else { return nil }
            return (.full, tightened)
        }
    }
}

File: android/app/src/main/java/com/jkpark/runwrap/engine/CollectionEngine.kt
package com.jkpark.runwrap.engine

import com.jkpark.runwrap.ui.Format
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.UUID
import kotlin.math.max

/// 도감 새 종류 (기획서 §5 도감) — 사이클의 목표 종목·기록이 어떤 새인지 결정한다
enum class BirdSpecies(val storageValue: String) {
    SPARROW("sparrow"), SWALLOW("swallow"), FALCON("falcon"),
    GOOSE("goose"), CRANE("crane"), SWAN("swan");

    val label: String
        get() = when (this) {
            SPARROW -> "참새"
            SWALLOW -> "제비"
            FALCON -> "매"
            GOOSE -> "기러기"
            CRANE -> "두루미"
            SWAN -> "백조"
        }

    /// 이 새를 얻는 목표 조건 안내 — 도감 카드·다음 목표 추천 문구
    val goalHint: String
        get() = when (this) {
            SPARROW -> "목표 없이 완주 습관"
            SWALLOW -> "5K · 10K"
            FALCON -> "하프"
            GOOSE -> "풀코스 완주"
            CRANE -> "풀코스 sub-4"
            SWAN -> "풀코스 서브3"
        }

    companion object {
        fun fromStorage(value: String): BirdSpecies? = entries.find { it.storageValue == value }
    }
}

/// 완성한 사이클 한 장 — 도감에 영구히 남는 기록.
/// 영속화(CollectionCache)는 스토어 계층 몫이다 (M5)
data class CollectedBird(
    val id: UUID = UUID.randomUUID(),
    val species: BirdSpecies,
    val goalLabel: String,     // "하프", "풀코스 3:59:00" 같은 그 사이클의 목표 표기
    val collectedAt: Instant,
    val cycleDays: Int,        // 사이클 시작부터 완성까지 걸린 일수
)

/// 도감 엔진 (기획서 §5) — 목표 종목·기록 → 새 종류 배정, 사이클 완성 판정,
/// 다음 목표 추천을 결정론적으로 계산한다. iOS `CollectionEngine.swift` 이식.
object CollectionEngine {
    /// 풀코스 기록 경계 (초) — 새 배정·목표 조이기 기준
    private const val SUB3 = 3 * 3_600
    private const val SUB4 = 4 * 3_600

    /// 다음 사이클 목표 추천 한 건
    data class GoalRecommendation(val distance: RaceDistance, val seconds: Int)

    /// 목표 종목·기록으로 이번 사이클의 새를 정한다.
    /// 목표 없음 → 참새, 5K·10K → 제비, 하프 → 매,
    /// 풀코스는 기록 없으면 기러기 · sub-4 두루미 · sub-3 백조
    fun species(distance: RaceDistance?, goalSeconds: Int): BirdSpecies {
        if (distance == null) return BirdSpecies.SPARROW
        return when (distance) {
            RaceDistance.FIVE_K, RaceDistance.TEN_K -> BirdSpecies.SWALLOW
            RaceDistance.HALF -> BirdSpecies.FALCON
            RaceDistance.FULL -> when {
                goalSeconds <= 0 -> BirdSpecies.GOOSE
                goalSeconds < SUB3 -> BirdSpecies.SWAN
                goalSeconds < SUB4 -> BirdSpecies.CRANE
                else -> BirdSpecies.GOOSE
            }
        }
    }

    /// 도감 카드에 적을 목표 표기 — 목표 없으면 참새 힌트, 기록 없으면 종목만
    fun goalLabel(distance: RaceDistance?, goalSeconds: Int): String {
        if (distance == null) return BirdSpecies.SPARROW.goalHint
        if (goalSeconds <= 0) return distance.label
        return "${distance.label} ${Format.duration(goalSeconds.toDouble())}"
    }

    /// 성체(최종 단계) 도달 = 사이클 완성 조건
    fun hasReachedAdult(stage: GrowthStage): Boolean = stage.next == null

    /// 사이클 완성 — 도감 카드 한 장을 만든다. 경과일은 자정 경계 기준
    fun collect(
        distance: RaceDistance?,
        goalSeconds: Int,
        cycleStartedAt: Instant,
        now: Instant,
        id: UUID = UUID.randomUUID(),
    ): CollectedBird {
        val days = ChronoUnit.DAYS.between(dayOf(cycleStartedAt), dayOf(now)).toInt()
        return CollectedBird(
            id = id,
            species = species(distance, goalSeconds),
            goalLabel = goalLabel(distance, goalSeconds),
            collectedAt = now,
            cycleDays = max(0, days),
        )
    }

    /// 다음 사이클 목표 추천 — 5K → 10K → 하프 → 풀 → sub-4 → 30분씩 조이기.
    /// 서브3 아래로는 더 조일 목표를 추천하지 않는다 (null)
    fun recommendedGoal(distance: RaceDistance?, goalSeconds: Int): GoalRecommendation? {
        if (distance == null) return GoalRecommendation(RaceDistance.FIVE_K, 0)
        return when (distance) {
            RaceDistance.FIVE_K -> GoalRecommendation(RaceDistance.TEN_K, 0)
            RaceDistance.TEN_K -> GoalRecommendation(RaceDistance.HALF, 0)
            RaceDistance.HALF -> GoalRecommendation(RaceDistance.FULL, 0)
            RaceDistance.FULL -> {
                if (goalSeconds <= 0) return GoalRecommendation(RaceDistance.FULL, SUB4)
                val tightened = goalSeconds - 30 * 60
                if (tightened < SUB3) null else GoalRecommendation(RaceDistance.FULL, tightened)
            }
        }
    }
}
